# Práctica 1: Aprendizaje Supervisado
## Dataset: Estimation of Obesity Levels Based on Eating Habits and Physical Condition

**Autor:** Jordi Blasco Lozano  
**DNI:** 74527208D  
**Email:** jbl42@alu.ua.es  
**Asignatura:** Aprendizaje Avanzado - Curso 2025/2026  
**Universidad de Alicante - Escuela Politécnica Superior**

> se reutiliza el mismo dataset y el mismo preprocesado de la Práctica 0.  

## 1. Carga de datos y configuración

En esta primera sección importamos todas las librerías necesarias, definimos funciones auxiliares para guardar tablas y figuras, y cargamos el dataset. Reutilizamos el mismo CSV de la Práctica 0 (*Estimation of Obesity Levels Based on Eating Habits and Physical Condition*) para garantizar comparabilidad directa con los resultados anteriores.

In [1]:
# Imports y configuración
from pathlib import Path
from time import perf_counter

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.pipeline import Pipeline

from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    classification_report, confusion_matrix, ConfusionMatrixDisplay
)

# Modelos
from sklearn.svm import SVC, LinearSVC
from sklearn.linear_model import SGDClassifier

from sklearn.tree import DecisionTreeClassifier, plot_tree, export_text
from sklearn.ensemble import (
    RandomForestClassifier, ExtraTreesClassifier,
    GradientBoostingClassifier, AdaBoostClassifier
)
from sklearn.inspection import permutation_importance

# Reproducibilidad
RANDOM_STATE = 42

# Carpetas de salida
BASE_DIR = Path('.').resolve()
IMAGES_DIR = BASE_DIR / 'images'
TABLES_DIR = BASE_DIR / 'tables'
IMAGES_DIR.mkdir(parents=True, exist_ok=True)
TABLES_DIR.mkdir(parents=True, exist_ok=True)

pd.set_option('display.max_columns', 200)
pd.set_option('display.width', 140)


In [2]:
def find_dataset_path() -> Path:
    """Busca el CSV en las rutas típicas del repo de prácticas."""
    candidates = [
        BASE_DIR / 'data' / 'ObesityDataSet_raw_and_data_sinthetic.csv',
        BASE_DIR.parent / 'practica_0' / 'data' / 'ObesityDataSet_raw_and_data_sinthetic.csv',
        BASE_DIR.parent / 'practica0' / 'data' / 'ObesityDataSet_raw_and_data_sinthetic.csv',
        BASE_DIR / 'ObesityDataSet_raw_and_data_sinthetic.csv',
    ]
    for p in candidates:
        if p.exists():
            return p
    raise FileNotFoundError(
        "No se encontró el dataset.\n\n"
        "Rutas probadas:\n - " + "\n - ".join(str(p) for p in candidates) + "\n\n"
        "Asegúrate de tener el CSV en practica_0/data o en practica_1/data."
    )

def save_table(df: pd.DataFrame, name: str) -> Path:
    """Guarda una tabla como LaTeX (solo tabular) en ./tables/{name}.tex"""
    out = TABLES_DIR / f"{name}.tex"
    latex = df.to_latex(index=False, escape=True, na_rep='-', float_format="%.4f", bold_rows=False)
    out.write_text(latex, encoding='utf-8')
    return out

def save_fig(path: Path, dpi: int = 300):
    plt.tight_layout()
    plt.savefig(path, dpi=dpi, bbox_inches='tight')
    plt.close()

def metrics_row(y_true, y_pred) -> dict:
    return {
        'Test Accuracy': accuracy_score(y_true, y_pred),
        'Precision (w)': precision_score(y_true, y_pred, average='weighted', zero_division=0),
        'Recall (w)': recall_score(y_true, y_pred, average='weighted', zero_division=0),
        'F1 (w)': f1_score(y_true, y_pred, average='weighted', zero_division=0),
    }

def plot_confusion(y_true, y_pred, class_names, filename: str, title: str):
    cm = confusion_matrix(y_true, y_pred)
    fig, ax = plt.subplots(figsize=(10, 8))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues',
                xticklabels=class_names, yticklabels=class_names, ax=ax)
    ax.set_xlabel('Predicho')
    ax.set_ylabel('Real')
    ax.set_title(title)
    ax.tick_params(axis='x', rotation=45)
    ax.tick_params(axis='y', rotation=0)
    save_fig(IMAGES_DIR / filename)



In [3]:
DATA_PATH = find_dataset_path()
df = pd.read_csv(DATA_PATH)
print('Ruta:', DATA_PATH)
print('Shape original:', df.shape)
df.head()


Ruta: /Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_0/data/ObesityDataSet_raw_and_data_sinthetic.csv
Shape original: (2111, 17)


,Gender,Age,Height,Weight,family_history_with_overweight,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,NObeyesdad
0,Female,21.0,1.62,64.0,yes,no,2.0,3.0,Sometimes,no,2.0,no,0.0,1.0,no,Public_Transportation,Normal_Weight
1,Female,21.0,1.52,56.0,yes,no,3.0,3.0,Sometimes,yes,3.0,yes,3.0,0.0,Sometimes,Public_Transportation,Normal_Weight
2,Male,23.0,1.80,77.0,yes,no,2.0,3.0,Sometimes,no,2.0,no,2.0,1.0,Frequently,Public_Transportation,Normal_Weight
3,Male,27.0,1.80,87.0,no,no,3.0,3.0,Sometimes,no,2.0,no,2.0,0.0,Frequently,Walking,Overweight_Level_I
4,Male,22.0,1.78,89.8,no,no,2.0,1.0,Sometimes,no,2.0,no,0.0,0.0,Sometimes,Public_Transportation,Overweight_Level_II


## 2. Preprocesado (idéntico a Práctica 0, salvo corrección de leakage en Box-Cox)

Tal como indica el enunciado, el preprocesado se da por supuesto y no es necesario detallarlo de nuevo. Para garantizar comparabilidad con la Práctica 0, se mantiene la misma lógica general de limpieza, con una corrección importante: la transformación **Box-Cox de `Age` se ajusta únicamente sobre el conjunto de entrenamiento** y después se aplica al conjunto de test usando el mismo valor de `lambda`, evitando así *data leakage*.

Los pasos aplicados quedan así:

1. **Eliminación de outliers en Weight y Height** mediante el método IQR (Rango Intercuartílico).
2. **Split train/test estratificado** 80/20 con `random_state=42`.
3. **Transformación Box-Cox en Age ajustada solo con train** y aplicada a test con el mismo `lambda`.
4. **Filtrado IQR post-Box-Cox sobre `Age` únicamente en train**, para que la limpieza también quede libre de fuga de información.
5. **Encoding de variables categóricas:** *Label Encoding* para las binarias (Gender, family_history, FAVC, SMOKE, SCC) y *One-Hot Encoding* con `drop_first=True` para las multiclase (CAEC, CALC, MTRANS), ajustado sobre train y alineado con test.
6. **Filtrado de outliers en train** con Z-score (|z| > 3) sobre las variables numéricas.

> **Nota sobre el escalado:** la estandarización no se aplica manualmente aquí, sino dentro de `Pipeline` (StandardScaler + modelo) para SVM y modelos lineales. Esto elimina el riesgo de *data leakage* y garantiza que en cada fold de validación cruzada la estandarización se recalcula de forma independiente.



In [4]:
target_col = 'NObeyesdad'

# 1) IQR en Weight y Height (se mantiene igual que en la práctica anterior)
df_clean = df.copy()
mask_clean = pd.Series(True, index=df_clean.index)
for col in ['Weight', 'Height']:
    q1 = df_clean[col].quantile(0.25)
    q3 = df_clean[col].quantile(0.75)
    iqr = q3 - q1
    lower = q1 - 1.5 * iqr
    upper = q3 + 1.5 * iqr
    mask_clean &= (df_clean[col] >= lower) & (df_clean[col] <= upper)

df_clean = df_clean.loc[mask_clean].reset_index(drop=True)

# 2) Split inicial ANTES de Box-Cox para evitar leakage en Age
X_raw = df_clean.drop(columns=[target_col]).copy()
y_raw = df_clean[target_col].copy()

X_train_raw, X_test_raw, y_train_raw, y_test_raw = train_test_split(
    X_raw, y_raw,
    test_size=0.2, random_state=RANDOM_STATE, stratify=y_raw
)

# 3) Box-Cox en Age ajustada solo en train y aplicada a test con el mismo lambda
X_train_raw = X_train_raw.copy()
X_test_raw = X_test_raw.copy()

X_train_raw['Age'], lambda_age = stats.boxcox(X_train_raw['Age'] + 1)
X_test_raw['Age'] = stats.boxcox(X_test_raw['Age'] + 1, lmbda=lambda_age)

# 4) Filtrado IQR post-Box-Cox solo en train
q1 = X_train_raw['Age'].quantile(0.25)
q3 = X_train_raw['Age'].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
mask_age_inliers = (X_train_raw['Age'] >= lower) & (X_train_raw['Age'] <= upper)

rows_before_age = len(X_train_raw)
X_train_raw = X_train_raw.loc[mask_age_inliers].reset_index(drop=True)
y_train_raw = y_train_raw.loc[mask_age_inliers].reset_index(drop=True)
rows_removed_age = rows_before_age - len(X_train_raw)

# 5) Encoding del target ajustado sobre train
le_target = LabelEncoder()
y_train = le_target.fit_transform(y_train_raw)
y_test = le_target.transform(y_test_raw)
class_names = le_target.classes_.tolist()

# 6) Encoding de features ajustado sobre train y alineado con test
X_train = X_train_raw.copy()
X_test = X_test_raw.copy()

binary_cols = ['Gender', 'family_history_with_overweight', 'FAVC', 'SMOKE', 'SCC']
for col in binary_cols:
    if col in X_train.columns:
        le = LabelEncoder()
        X_train[col] = le.fit_transform(X_train[col])
        X_test[col] = le.transform(X_test[col])

multi_cat_cols = ['CAEC', 'CALC', 'MTRANS']
X_train = pd.get_dummies(X_train, columns=multi_cat_cols, drop_first=True)
X_test = pd.get_dummies(X_test, columns=multi_cat_cols, drop_first=True)
X_test = X_test.reindex(columns=X_train.columns, fill_value=0)

# Columnas numéricas para Z-score post-split
numeric_cols = X_train_raw.select_dtypes(include=[np.number]).columns.tolist()

# 7) Filtrado adicional de outliers en train con Z-score
scaler_temp = StandardScaler()
X_train_z = scaler_temp.fit_transform(X_train[numeric_cols])
mask_inliers = (np.abs(X_train_z) <= 3).all(axis=1)

rows_before_z = len(X_train)
X_train = X_train.loc[mask_inliers].reset_index(drop=True)
y_train = y_train[mask_inliers]
rows_removed_z = rows_before_z - len(X_train)

print('Shape tras limpieza Weight/Height:', df_clean.shape)
print(f'Lambda Box-Cox (ajustado en train): {lambda_age:.4f}')
print(f'Filas eliminadas en train por IQR tras Box-Cox (Age): {rows_removed_age}')
print('Shape tras encoding train/test:', X_train.shape, X_test.shape)
print(f'Outliers eliminados en train por Z-score: {rows_removed_z}')



Shape tras limpieza Weight/Height: (2109, 17)
Lambda Box-Cox (ajustado en train): -1.6770
Filas eliminadas en train por IQR tras Box-Cox (Age): 1
Shape tras encoding train/test: (1686, 23) (422, 23)
Outliers eliminados en train por Z-score: 0


## 3. Parte 1 — Support Vector Machines (SVM)

Las SVM buscan el hiperplano que maximiza el margen entre clases. En problemas no linealmente separables, utilizan funciones *kernel* para proyectar los datos a un espacio de mayor dimensionalidad donde sí son separables. Los principales hiperparámetros son:

- **C:** controla el trade-off entre maximizar el margen y minimizar errores de clasificación. Un C alto genera un margen estrecho con pocos errores (riesgo de *overfitting*), mientras que un C bajo genera un margen amplio pero permite más errores (*underfitting*).
- **gamma (γ):** en kernels RBF y polinomial, controla el alcance de influencia de cada muestra. Un gamma alto implica influencia local (fronteras complejas), y un gamma bajo implica influencia amplia (fronteras suaves).
- **degree:** grado del polinomio en el kernel polinomial.


### 3.1 SVM con diferentes kernels (parámetros por defecto)

Empezamos entrenando modelos SVM con los cuatro kernels principales (Linear, Polinomial grado 3, RBF y Sigmoid) usando los hiperparámetros por defecto de scikit-learn. Para cada kernel medimos: accuracy en validación cruzada 5-Fold, accuracy y métricas ponderadas en test, número de vectores de soporte y tiempo de entrenamiento.

In [5]:
kernels = {
    'Linear':     SVC(kernel='linear', random_state=RANDOM_STATE),
    'Poly (d=3)': SVC(kernel='poly', degree=3, random_state=RANDOM_STATE),
    'RBF':        SVC(kernel='rbf', random_state=RANDOM_STATE),
    'Sigmoid':    SVC(kernel='sigmoid', random_state=RANDOM_STATE),
}

rows = []
for name, svc in kernels.items():
    pipe = Pipeline([('scaler', StandardScaler()), ('svm', svc)])

    # CV
    cv_scores = cross_val_score(pipe, X_train, y_train, cv=5, scoring='accuracy')
    cv_mean, cv_std = cv_scores.mean(), cv_scores.std()

    # Fit + test
    t0 = perf_counter()
    pipe.fit(X_train, y_train)
    fit_time = perf_counter() - t0

    y_pred = pipe.predict(X_test)
    m = metrics_row(y_test, y_pred)

    n_sv = int(pipe.named_steps['svm'].n_support_.sum())
    rows.append({
        'Kernel': name,
        'CV Accuracy': cv_mean,
        **m,
        'Nº SV': n_sv,
        '% Train SV': 100 * n_sv / len(X_train),
        'Fit Time (s)': fit_time,
    })

df_svm_kernels = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)

# Tabla con columnas EXACTAS del enunciado/memoria
df_svm_kernels_memoria = (
    df_svm_kernels
    .rename(columns={
        'Precision (w)': 'Precision',
        'Recall (w)': 'Recall',
        'F1 (w)': 'F1-Score',
        'Nº SV': 'N° SV',
        '% Train SV': '% Training'
    })
    [['Kernel', 'CV Accuracy', 'Test Accuracy', 'Precision', 'Recall', 'F1-Score', 'N° SV', '% Training']]
)

df_svm_kernels_memoria


,Kernel,CV Accuracy,Test Accuracy,Precision,Recall,F1-Score,N° SV,% Training
0,Linear,0.949576,0.943128,0.943759,0.943128,0.943072,564,33.451957
2,RBF,0.857648,0.874408,0.879555,0.874408,0.875817,1080,64.056940
1,Poly (d=3),0.767503,0.739336,0.749586,0.739336,0.732298,1203,71.352313
3,Sigmoid,0.645299,0.642180,0.640221,0.642180,0.640470,1222,72.479241


El kernel **Linear** vuelve a obtener el mejor resultado con parámetros por defecto, y además mejora ligeramente respecto a la versión anterior del notebook: **CV = 0.9496** y **Test = 0.9431**. Esto refuerza la idea de que, tras el preprocesado y el encoding, el problema queda representado en un espacio de 23 dimensiones donde gran parte de la separación entre clases puede capturarse con una frontera lineal. La corrección del *data leakage* en `Age` no perjudica al kernel lineal; de hecho, lo deja como la referencia más sólida dentro de SVM.

El kernel **RBF** por defecto mejora algo respecto a la ejecución previa (**CV = 0.8576**, **Test = 0.8744**), pero sigue muy por debajo del lineal si no se ajustan `C` y `gamma`. Esto confirma que RBF necesita sintonización cuidadosa para desplegar su potencial.

El kernel **Sigmoid** vuelve a ser el peor con diferencia (**CV = 0.6453**, **Test = 0.6422**), lo que encaja con su comportamiento habitual: es mucho menos estable que los otros kernels y tiende a depender de una parametrización bastante fina.

También es interesante observar el número de **vectores de soporte (SV)**: el kernel lineal usa **564 SV** (33.45 % del training set), mientras que los kernels no lineales necesitan entre el **64 % y el 72 %**. A mayor número de SV, mayor complejidad efectiva del clasificador y mayor coste en inferencia.


In [6]:
save_table(df_svm_kernels_memoria, 'svm_kernels')


PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_kernels.tex')

### 3.2 Optimización del kernel RBF (GridSearchCV)

El kernel RBF (`exp(-γ·||x-x'||²)`) suele ser el más efectivo para datos no lineales, pero depende fuertemente de dos hiperparámetros:

- **C** (regularización): valores altos → margen pequeño, se ajusta más a los datos.
- **gamma (γ)**: valores altos → cada muestra influye solo en su vecindad cercana, produciendo fronteras más complejas.

Utilizamos `GridSearchCV` con CV 5-Fold para explorar combinaciones de C ∈ {0.1, 1, 10, 100} y γ ∈ {0.001, 0.01, 0.1, 1, 'scale', 'auto'}.

In [7]:
pipe_rbf = Pipeline([
    ('scaler', StandardScaler()),
    ('svm', SVC(kernel='rbf', random_state=RANDOM_STATE))
])

param_grid_rbf = {
    'svm__C': [0.1, 1, 10, 100],
    'svm__gamma': [0.001, 0.01, 0.1, 1, 'scale', 'auto']
}

grid_rbf = GridSearchCV(
    pipe_rbf, param_grid_rbf,
    cv=5, scoring='accuracy', n_jobs=-1, refit=True
)

t0 = perf_counter()
grid_rbf.fit(X_train, y_train)
grid_time = perf_counter() - t0

print("Mejores parámetros:", grid_rbf.best_params_)
print(f"Mejor CV score: {grid_rbf.best_score_:.4f}")
print(f"Test score: {grid_rbf.best_estimator_.score(X_test, y_test):.4f}")
print(f"Tiempo GridSearch: {grid_time:.3f}s")


Mejores parámetros: {'svm__C': 100, 'svm__gamma': 0.01}
Mejor CV score: 0.9401
Test score: 0.9408
Tiempo GridSearch: 10.158s


In [8]:
# Tabla con al menos 10 combinaciones representativas: usamos las 10 mejores por CV
cvres = pd.DataFrame(grid_rbf.cv_results_)
cvres = cvres.sort_values('mean_test_score', ascending=False)

top = cvres.head(10).copy()
rows = []
for _, r in top.iterrows():
    C = r['param_svm__C']
    gamma = r['param_svm__gamma']

    pipe = Pipeline([
        ('scaler', StandardScaler()),
        ('svm', SVC(kernel='rbf', C=C, gamma=gamma, random_state=RANDOM_STATE))
    ])
    t0 = perf_counter()
    pipe.fit(X_train, y_train)
    t_fit = perf_counter() - t0

    y_pred = pipe.predict(X_test)
    n_sv = int(pipe.named_steps['svm'].n_support_.sum())
    rows.append({
        'C': float(C),
        'gamma': str(gamma),
        'CV Accuracy': float(r['mean_test_score']),
        'Test Accuracy': accuracy_score(y_test, y_pred),
        'Nº SV': n_sv,
        '% Train': 100 * n_sv / len(X_train),
        'Tiempo (s)': t_fit
    })

df_svm_rbf = pd.DataFrame(rows)
df_svm_rbf


,C,gamma,CV Accuracy,Test Accuracy,Nº SV,% Train,Tiempo (s)
0,100.0,0.01,0.940087,0.940758,632,37.485172,0.067595
1,100.0,0.001,0.928218,0.928910,868,51.482800,0.064135
2,10.0,0.01,0.911613,0.924171,874,51.838671,0.063342
3,100.0,auto,0.910428,0.917062,814,48.279953,0.074091
4,100.0,scale,0.910428,0.917062,814,48.279953,0.078711
5,10.0,scale,0.905095,0.902844,858,50.889680,0.074215
6,10.0,auto,0.905095,0.902844,858,50.889680,0.074306
7,10.0,0.1,0.894422,0.890995,954,56.583630,0.103734
8,100.0,0.1,0.894416,0.898104,941,55.812574,0.090544
9,1.0,0.1,0.875443,0.867299,1043,61.862396,0.084844


In [9]:
save_table(df_svm_rbf, 'svm_rbf_grid')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_rbf_grid.tex')

La mejor combinación encontrada sigue siendo **C = 100** y **γ = 0.01**, pero con la nueva ejecución libre de leakage el resultado queda en **CV = 0.9401** y **Test = 0.9408**. La mejora respecto al RBF por defecto sigue siendo muy clara, aunque ahora el RBF optimizado ya no supera al kernel lineal: se queda ligeramente por debajo tanto en validación cruzada como en test.

Este resultado es interesante porque sugiere que la pequeña ventaja aparente del RBF en la versión anterior probablemente estaba inflada por el preprocesado de `Age` sobre todo el dataset. Con la corrección metodológica, la no linealidad sigue siendo útil, pero ya no parece imprescindible. El valor alto de `C` indica que el modelo se beneficia de penalizar más los errores, mientras que `γ = 0.01` mantiene una influencia moderada de cada muestra y evita una frontera excesivamente irregular.

### 3.3 Optimización del kernel polinomial (GridSearchCV)

El kernel polinomial (`(γ·⟨x, x'⟩ + r)^d`) permite controlar la complejidad de la frontera mediante el grado (degree) y la combinación de C y γ. Exploramos degree ∈ {2, 3, 4}, C ∈ {0.1, 1, 10} y γ ∈ {'scale', 'auto', 0.1}.


In [10]:
pipe_poly = Pipeline([
    ('scaler', StandardScaler()),
    ('svm', SVC(kernel='poly', random_state=RANDOM_STATE))
])

param_grid_poly = {
    'svm__degree': [2, 3, 4],
    'svm__C': [0.1, 1, 10],
    'svm__gamma': ['scale', 'auto', 0.1]
}

grid_poly = GridSearchCV(
    pipe_poly, param_grid_poly,
    cv=5, scoring='accuracy', n_jobs=-1, refit=True
)

t0 = perf_counter()
grid_poly.fit(X_train, y_train)
grid_time_poly = perf_counter() - t0

print("Mejores parámetros:", grid_poly.best_params_)
print(f"Mejor CV score: {grid_poly.best_score_:.4f}")
print(f"Test score: {grid_poly.best_estimator_.score(X_test, y_test):.4f}")
print(f"Tiempo GridSearch: {grid_time_poly:.3f}s")


Mejores parámetros: {'svm__C': 10, 'svm__degree': 3, 'svm__gamma': 0.1}
Mejor CV score: 0.8808
Test score: 0.8791
Tiempo GridSearch: 1.411s


In [11]:
cvres_poly = pd.DataFrame(grid_poly.cv_results_).sort_values('mean_test_score', ascending=False)

# Cogemos 10 representativas (top 10)
top = cvres_poly.head(10).copy()
rows = []
for _, r in top.iterrows():
    degree = int(r['param_svm__degree'])
    C = float(r['param_svm__C'])
    gamma = r['param_svm__gamma']

    pipe = Pipeline([
        ('scaler', StandardScaler()),
        ('svm', SVC(kernel='poly', degree=degree, C=C, gamma=gamma, random_state=RANDOM_STATE))
    ])
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)

    n_sv = int(pipe.named_steps['svm'].n_support_.sum())
    rows.append({
        'degree': degree,
        'C': C,
        'gamma': str(gamma),
        'CV Accuracy': float(r['mean_test_score']),
        'Test Accuracy': accuracy_score(y_test, y_pred),
        'Nº SV': n_sv
    })

df_svm_poly = pd.DataFrame(rows)
df_svm_poly


,degree,C,gamma,CV Accuracy,Test Accuracy,Nº SV
0,3,10.0,0.1,0.880786,0.879147,792
1,3,1.0,0.1,0.873066,0.862559,892
2,3,10.0,auto,0.867728,0.864929,909
3,3,10.0,scale,0.867728,0.864929,909
4,2,10.0,0.1,0.857044,0.834123,781
5,2,10.0,auto,0.853495,0.841232,898
6,2,10.0,scale,0.853495,0.841232,898
7,2,1.0,0.1,0.844593,0.841232,957
8,4,10.0,0.1,0.842814,0.845972,862
9,4,1.0,0.1,0.840435,0.845972,962


In [12]:
save_table(df_svm_poly, 'svm_poly_grid')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_poly_grid.tex')

La mejor configuración polinomial pasa a ser **degree = 3, C = 10, γ = 0.1** con **CV = 0.8808** y **Test = 0.8791**. El rendimiento mejora frente a la ejecución anterior, pero sigue claramente por debajo tanto del kernel lineal como del RBF optimizado. Grados mayores (`d = 4`) no aportan mejora, lo que sugiere que el problema no necesita interacciones polinómicas de orden alto para modelarse bien.

### 3.4 Comparación de implementaciones lineales (SVC linear vs LinearSVC vs SGDClassifier)

Scikit-learn ofrece tres implementaciones del caso lineal con distinta formulación interna y coste computacional:

| Implementación | Formulación | Uso recomendado |
|:---|:---|:---|
| `SVC(kernel='linear')` | Dual (libsvm) | n_samples < 10,000 |
| `LinearSVC` | Primal (liblinear) | Datasets medianos, kernel lineal |
| `SGDClassifier(loss='hinge')` | SGD con hinge loss | Datasets grandes (> 100,000) |

Comparamos rendimiento y tiempo de entrenamiento de las tres implementaciones.


In [13]:
models_linear = {
    'SVC(linear)':   Pipeline([('scaler', StandardScaler()),
                               ('svm', SVC(kernel='linear', random_state=RANDOM_STATE))]),
    'LinearSVC':     Pipeline([('scaler', StandardScaler()),
                               ('svm', LinearSVC(random_state=RANDOM_STATE, max_iter=10000))]),
    'SGDClassifier': Pipeline([('scaler', StandardScaler()),
                               ('svm', SGDClassifier(loss='hinge', random_state=RANDOM_STATE))]),
}

rows = []
for name, model in models_linear.items():
    t0 = perf_counter()
    model.fit(X_train, y_train)
    fit_time = perf_counter() - t0

    cv = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
    y_pred = model.predict(X_test)

    n_sv = '-'
    if name.startswith('SVC'):
        n_sv = int(model.named_steps['svm'].n_support_.sum())

    rows.append({
        'Implementación': name,
        'CV Accuracy': cv.mean(),
        'Test Accuracy': accuracy_score(y_test, y_pred),
        'Nº SV': n_sv,
        'Tiempo (s)': fit_time
    })

df_svm_linear_impl = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
df_svm_linear_impl


,Implementación,CV Accuracy,Test Accuracy,Nº SV,Tiempo (s)
0,SVC(linear),0.949576,0.943128,564,0.064340
1,LinearSVC,0.772841,0.779621,-,0.031206
2,SGDClassifier,0.719456,0.680095,-,0.042705


In [14]:
save_table(df_svm_linear_impl, 'svm_linear_impl')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_linear_impl.tex')

`SVC(linear)` sigue siendo claramente superior, y ahora incluso mejora sus cifras: **CV = 0.9496** y **Test = 0.9431**. `LinearSVC` queda bastante por detrás (**CV = 0.7728**, **Test = 0.7796**) y `SGDClassifier` todavía más abajo (**CV = 0.7195**, **Test = 0.6801**). La diferencia se explica porque `SVC` resuelve el problema dual exacto, mientras que `LinearSVC` usa otra formulación y `SGDClassifier` aproxima la solución mediante optimización estocástica, más sensible a hiperparámetros y convergencia.

En tiempo, las tres implementaciones siguen siendo muy rápidas dado el tamaño reducido del dataset. Para este problema, la ligera ventaja de velocidad de `LinearSVC` o `SGDClassifier` no compensa la pérdida de precisión.

### 3.5 Análisis de vectores de soporte (configuraciones representativas)

Los vectores de soporte son las muestras de entrenamiento que definen la frontera de decisión. A mayor número de SV, el modelo utiliza más datos para trazar la frontera, lo que puede indicar mayor complejidad o mayor dificultad del problema para ese kernel. Comparamos el número de SV en distintas configuraciones.


In [15]:
configs = [
    ('Linear (C=1)',        dict(kernel='linear', C=1, gamma='scale')),
    ('RBF (C=1, γ=scale)',  dict(kernel='rbf', C=1, gamma='scale')),
    ('RBF (C=10, γ=0.1)',   dict(kernel='rbf', C=10, gamma=0.1)),
    ('RBF (C=100, γ=1)',    dict(kernel='rbf', C=100, gamma=1)),
    ('Poly (d=3, C=1)',     dict(kernel='poly', degree=3, C=1, gamma='scale')),
]

rows = []
for name, params in configs:
    svc = SVC(random_state=RANDOM_STATE, **params)
    pipe = Pipeline([('scaler', StandardScaler()), ('svm', svc)])
    pipe.fit(X_train, y_train)

    n_sv = int(pipe.named_steps['svm'].n_support_.sum())
    rows.append({
        'Configuración': name,
        'Nº SV': n_sv,
        '% Training Set': 100 * n_sv / len(X_train),
    })

df_sv = pd.DataFrame(rows).sort_values('% Training Set')
df_sv


,Configuración,Nº SV,% Training Set
0,Linear (C=1),564,33.451957
2,"RBF (C=10, γ=0.1)",954,56.583630
1,"RBF (C=1, γ=scale)",1080,64.056940
4,"Poly (d=3, C=1)",1203,71.352313
3,"RBF (C=100, γ=1)",1268,75.207592


In [16]:
save_table(df_sv, 'svm_support_vectors')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_support_vectors.tex')

La configuración con menor porcentaje de SV es el kernel lineal (C = 1), lo que indica que necesita relativamente pocos puntos de referencia para definir la frontera. La configuración extrema **RBF (C = 100, γ = 1)** usa una proporción elevada del training set como SV, señal de que con γ alto la frontera es tan local que prácticamente memoriza los datos (riesgo de *overfitting*).

### 3.6 Evaluación final SVM (mejor modelo) + matriz de confusión

Evaluamos el mejor modelo SVM encontrado (RBF con C = 100, γ = 0.01) en el conjunto de test y generamos su *classification report* y matriz de confusión.

In [17]:
best_svm = grid_rbf.best_estimator_
best_svm.fit(X_train, y_train)
y_pred_svm = best_svm.predict(X_test)

print(classification_report(y_test, y_pred_svm, target_names=class_names))

plot_confusion(
    y_test, y_pred_svm, class_names,
    filename='confusion_matrix_svm.png',
    title='Matriz de Confusión — Mejor SVM'
)


                     precision    recall  f1-score   support

Insufficient_Weight       0.95      1.00      0.97        54
      Normal_Weight       0.89      0.82      0.85        57
     Obesity_Type_I       0.92      0.99      0.95        70
    Obesity_Type_II       0.98      0.93      0.96        60
   Obesity_Type_III       1.00      0.98      0.99        65
 Overweight_Level_I       0.87      0.91      0.89        58
Overweight_Level_II       0.98      0.93      0.96        58

           accuracy                           0.94       422
          macro avg       0.94      0.94      0.94       422
       weighted avg       0.94      0.94      0.94       422



In [18]:
# Tabla resumen (Parte 1)
def summarize_model(name, model, is_svm=True):
    cv = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    m = metrics_row(y_test, y_pred)
    n_sv = '-'
    pct = '-'
    if is_svm and hasattr(model.named_steps['svm'], 'n_support_'):
        n_sv = int(model.named_steps['svm'].n_support_.sum())
        pct = 100 * n_sv / len(X_train)
    return {
        'Modelo': name,
        'CV Accuracy': cv.mean(),
        'Test Accuracy': m['Test Accuracy'],
        'F1 (w)': m['F1 (w)'],
        'Nº SV': n_sv,
        '% Train SV': pct
    }

svm_summary_rows = [
    summarize_model('SVM Linear', Pipeline([('scaler', StandardScaler()),
                                           ('svm', SVC(kernel='linear', random_state=RANDOM_STATE))]), True),
    summarize_model('SVM Poly (best)', grid_poly.best_estimator_, True),
    summarize_model('SVM RBF (best)', grid_rbf.best_estimator_, True),
    summarize_model('SVM Sigmoid', Pipeline([('scaler', StandardScaler()),
                                            ('svm', SVC(kernel='sigmoid', random_state=RANDOM_STATE))]), True),
]

df_svm_summary = pd.DataFrame(svm_summary_rows).sort_values('CV Accuracy', ascending=False)
df_svm_summary


,Modelo,CV Accuracy,Test Accuracy,F1 (w),Nº SV,% Train SV
0,SVM Linear,0.949576,0.943128,0.943072,564,33.451957
2,SVM RBF (best),0.940087,0.940758,0.940521,632,37.485172
1,SVM Poly (best),0.880786,0.879147,0.877790,792,46.975089
3,SVM Sigmoid,0.645299,0.642180,0.640470,1222,72.479241


In [19]:
save_table(df_svm_summary, 'svm_summary')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/svm_summary.tex')

El mejor modelo no lineal dentro de la familia sigue siendo **SVM-RBF** con **C = 100** y **γ = 0.01**, alcanzando **Test Accuracy = 0.9408** y **F1 ponderado = 0.9405**. Sin embargo, el resultado más importante tras corregir el leakage es que el **kernel lineal pasa a ser ligeramente mejor** (**0.9431** en test frente a **0.9408** del RBF). Esto indica que el problema es incluso más lineal de lo que parecía en la versión anterior.

Las clases extremas siguen obteniendo F1 muy altos, mientras que las mayores confusiones aparecen en **Normal_Weight** y **Overweight_Level_I**, que son categorías contiguas y más difíciles de separar. La tabla resumen de la Parte 1 recoge los mejores resultados por cada familia de kernel.

---

## 4. Parte 2 — Árboles de Decisión

Los árboles de decisión (CART) realizan particiones recursivas del espacio de features según criterios de impureza (Gini o entropía). Su principal ventaja es la interpretabilidad: producen reglas de decisión legibles. El reto principal es controlar el *overfitting*, ya que sin restricciones un árbol puede memorizar completamente el training set. Para combatirlo se usan dos estrategias:

- **Poda previa (pre-pruning):** limitar la profundidad (`max_depth`), el mínimo de muestras por split/hoja o el número de hojas.
- **Poda posterior (post-pruning, CCP):** entrenar el árbol completo y después ir podando ramas usando el parámetro `ccp_alpha` (cost-complexity pruning).


### 4.1 Árbol sin restricciones (baseline)

Empezamos por entrenar un árbol sin ninguna restricción para observar el comportamiento de *overfitting*: esperamos que alcance un 100% de accuracy en train pero un rendimiento inferior en test.

In [20]:
tree_full = DecisionTreeClassifier(random_state=RANDOM_STATE)
t0 = perf_counter()
tree_full.fit(X_train, y_train)
t_fit = perf_counter() - t0

train_acc = tree_full.score(X_train, y_train)
test_acc = tree_full.score(X_test, y_test)

df_tree_baseline = pd.DataFrame([{
    'Configuración': 'Sin restricciones',
    'Train Accuracy': train_acc,
    'Test Accuracy': test_acc,
    'Profundidad': tree_full.get_depth(),
    'Nº Hojas': tree_full.get_n_leaves(),
    'Tiempo (s)': t_fit
}])
df_tree_baseline


,Configuración,Train Accuracy,Test Accuracy,Profundidad,Nº Hojas,Tiempo (s)
0,Sin restricciones,1.0,0.92654,11,97,0.011658


In [21]:
save_table(df_tree_baseline, 'tree_baseline')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_baseline.tex')

Efectivamente, el árbol sin restricciones alcanza **100 % en train** pero solo **92.65 % en test**, con una profundidad real de **11 niveles** y **97 hojas**. La brecha entre train y test sigue siendo la señal clásica de *overfitting*: el árbol es capaz de memorizar el entrenamiento, pero esa complejidad adicional no se traduce en una mejora de generalización.

### 4.2 Efecto de max_depth

Exploramos cómo limitar la profundidad máxima del árbol afecta al equilibrio entre sesgo y varianza. A poca profundidad (1–3), el modelo es demasiado simple (*underfitting*); a profundidad sin límite, se produce *overfitting*. El objetivo es encontrar el punto intermedio.


In [22]:
depths = [1, 2, 3, 5, 7, 10, 15, None]
rows = []

for depth in depths:
    tree = DecisionTreeClassifier(max_depth=depth, random_state=RANDOM_STATE)

    cv_scores = cross_val_score(tree, X_train, y_train, cv=5, scoring='accuracy')
    cv_mean = cv_scores.mean()

    tree.fit(X_train, y_train)
    rows.append({
        'max_depth': 'None' if depth is None else depth,
        'Train Acc': tree.score(X_train, y_train),
        'CV Acc': cv_mean,
        'Test Acc': tree.score(X_test, y_test),
        'Profundidad real': tree.get_depth(),
        'Nº Hojas': tree.get_n_leaves()
    })

df_tree_depth = pd.DataFrame(rows)
df_tree_depth


,max_depth,Train Acc,CV Acc,Test Acc,Profundidad real,Nº Hojas
0,1,0.290629,0.290629,0.291469,1,2
1,2,0.553974,0.552786,0.542654,2,4
2,3,0.643535,0.634037,0.649289,3,8
3,5,0.852313,0.831545,0.817536,5,25
4,7,0.947212,0.893219,0.905213,7,57
5,10,0.998814,0.934744,0.926540,10,95
6,15,1.000000,0.938899,0.926540,11,97
7,None,1.000000,0.938305,0.926540,11,97


In [23]:
# Curva profundidad vs rendimiento
dfp = df_tree_depth.copy()
dfp['max_depth_num'] = dfp['max_depth'].replace('None', np.nan).astype(float)

fig, ax = plt.subplots(figsize=(10, 5))
ax.plot(dfp['max_depth_num'], dfp['Train Acc'], marker='o', label='Train')
ax.plot(dfp['max_depth_num'], dfp['CV Acc'], marker='o', label='CV')
ax.plot(dfp['max_depth_num'], dfp['Test Acc'], marker='o', label='Test')
ax.set_xlabel('max_depth (None omitido)')
ax.set_ylabel('Accuracy')
ax.set_title('Efecto de max_depth en Decision Tree')
ax.grid(True, alpha=0.3)
ax.legend()
save_fig(IMAGES_DIR / 'decision_tree_depth_curve.png')


/var/folders/0m/q5gmrh695pvc0bcq9s0zmcqh0000gr/T/ipykernel_12704/1536526240.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dfp['max_depth_num'] = dfp['max_depth'].replace('None', np.nan).astype(float)


In [24]:
save_table(df_tree_depth, 'tree_max_depth')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_max_depth.tex')

La gráfica y la tabla vuelven a mostrar el patrón clásico. Con **max_depth = 1** el árbol apenas puede separar dos grandes regiones y se queda en torno al **29 %** de accuracy. A medida que aumenta la profundidad, el rendimiento mejora de forma sostenida hasta estabilizarse en torno a **10–15 niveles**. A partir de ahí, el train prácticamente se hace perfecto, mientras que CV y test dejan de mejorar.

Es interesante notar que la validación cruzada máxima de esta exploración simple aparece en **max_depth = 15** (**CV = 0.9389**), pero el test no mejora respecto a profundidad 10 o al árbol completo. Es la señal de que, una vez capturadas las reglas principales, seguir profundizando aporta más ajuste fino que capacidad real de generalización.

### 4.3 Criterios de división

Comparamos los tres criterios de impureza disponibles para clasificación: **Gini** (usado por defecto en CART), **Entropía** (ID3/C4.5) y **Log Loss**. Los tres miden la heterogeneidad de los nodos, pero con fórmulas ligeramente distintas.


In [25]:
criterios = ['gini', 'entropy', 'log_loss']
rows = []
for crit in criterios:
    tree = DecisionTreeClassifier(criterion=crit, random_state=RANDOM_STATE)
    cv_scores = cross_val_score(tree, X_train, y_train, cv=5, scoring='accuracy')

    tree.fit(X_train, y_train)
    rows.append({
        'Criterio': crit,
        'CV Accuracy': cv_scores.mean(),
        'CV Std': cv_scores.std(),
        'Test Accuracy': tree.score(X_test, y_test),
        'Nº Hojas': tree.get_n_leaves()
    })

df_tree_criteria = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
df_tree_criteria


,Criterio,CV Accuracy,CV Std,Test Accuracy,Nº Hojas
1,entropy,0.943053,0.013082,0.973934,81
2,log_loss,0.943053,0.013082,0.973934,81
0,gini,0.938305,0.010725,0.926540,97


In [26]:
save_table(df_tree_criteria[['Criterio','CV Accuracy','Test Accuracy','Nº Hojas']], 'tree_criteria')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_criteria.tex')

Con la nueva ejecución, el cambio de criterio deja de ser un detalle menor y pasa a ser uno de los resultados más llamativos de toda la práctica. Los criterios **entropy** y **log_loss** obtienen **CV = 0.9431** y un **Test Accuracy = 0.9739**, mientras que **gini** se queda en **0.9265** en test. Además, entropy/log_loss lo logran con **81 hojas**, frente a las **97** de gini.

Esto significa que, en este dataset y con el nuevo preprocesado sin leakage, el criterio de división sí importa mucho. La entropía está encontrando splits más informativos y consigue un árbol individual sorprendentemente competitivo, hasta el punto de rivalizar con los mejores ensembles. Es un hallazgo importante porque demuestra que, antes de asumir que un árbol individual tiene un techo bajo, conviene revisar cuidadosamente cómo se está construyendo.

### 4.4 Poda previa — GridSearchCV

Además de `max_depth`, existen otros parámetros de poda previa: `min_samples_split` (mínimo de muestras para dividir un nodo), `min_samples_leaf` (mínimo de muestras en cada hoja) y `max_leaf_nodes` (máximo número de hojas). Exploramos combinaciones de todos ellos con `GridSearchCV`.


In [27]:
param_grid_tree = {
    'max_depth':         [3, 5, 10, None],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf':  [1, 2, 5, 10],
    'max_leaf_nodes':    [None, 10, 20, 50]
}

grid_tree = GridSearchCV(
    DecisionTreeClassifier(random_state=RANDOM_STATE),
    param_grid_tree,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    refit=True
)

t0 = perf_counter()
grid_tree.fit(X_train, y_train)
t_grid = perf_counter() - t0

print("Mejores parámetros:", grid_tree.best_params_)
print(f"Mejor CV score: {grid_tree.best_score_:.4f}")
print(f"Test score: {grid_tree.best_estimator_.score(X_test, y_test):.4f}")
print(f"Tiempo GridSearch: {t_grid:.3f}s")


Mejores parámetros: {'max_depth': None, 'max_leaf_nodes': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Mejor CV score: 0.9383
Test score: 0.9265
Tiempo GridSearch: 1.430s


In [28]:
cvres_tree = pd.DataFrame(grid_tree.cv_results_).sort_values('mean_test_score', ascending=False)
top = cvres_tree.head(10).copy()

rows = []
for _, r in top.iterrows():
    params = {
        'max_depth': r['param_max_depth'],
        'min_samples_split': int(r['param_min_samples_split']),
        'min_samples_leaf': int(r['param_min_samples_leaf']),
        'max_leaf_nodes': r['param_max_leaf_nodes'],
    }
    model = DecisionTreeClassifier(random_state=RANDOM_STATE, **params)
    model.fit(X_train, y_train)
    rows.append({
        'max_depth': 'None' if params['max_depth'] is None else int(params['max_depth']),
        'min_samples_split': params['min_samples_split'],
        'min_samples_leaf': params['min_samples_leaf'],
        'max_leaf_nodes': 'None' if params['max_leaf_nodes'] is None else int(params['max_leaf_nodes']),
        'CV Acc': float(r['mean_test_score']),
        'Test Acc': model.score(X_test, y_test),
    })

df_tree_grid = pd.DataFrame(rows)
df_tree_grid


,max_depth,min_samples_split,min_samples_leaf,max_leaf_nodes,CV Acc,Test Acc
0,None,2,1,None,0.938305,0.926540
1,None,5,1,None,0.935933,0.928910
2,10,2,1,None,0.934744,0.926540
3,10,5,1,None,0.934153,0.928910
4,None,2,2,50,0.930588,0.936019
5,None,2,2,None,0.929405,0.928910
6,None,5,2,50,0.928808,0.936019
7,None,10,1,None,0.928222,0.912322
8,None,5,2,None,0.928220,0.924171
9,10,2,2,50,0.928214,0.936019


In [29]:
save_table(df_tree_grid, 'tree_grid_prepruning')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_grid_prepruning.tex')

La mejor configuración de poda previa encontrada por GridSearch pasa a ser simplemente el árbol sin restricciones dentro del espacio explorado: **max_depth = None, min_samples_leaf = 1, min_samples_split = 2, max_leaf_nodes = None** con **CV = 0.9383** y **Test = 0.9265**. Es decir, en esta búsqueda concreta la poda previa no aporta mejora real frente al baseline.

Aun así, aparecen configuraciones algo más restringidas con **Test = 0.9360** y CV ligeramente inferior, lo que sugiere que una poda suave puede mejorar el comportamiento en test aunque no sea la ganadora estricta por validación cruzada. También conviene recordar que este grid no está variando el criterio de división, y precisamente ahí es donde hemos visto el mayor salto de calidad.

### 4.5 Poda posterior — Cost-Complexity Pruning (ccp_alpha)

CART implementa poda posterior mediante el parámetro `ccp_alpha`. A mayor valor de alpha, mayor poda: se eliminan las ramas cuya contribución a la reducción de impureza no compense el incremento de complejidad. El procedimiento es: (1) entrenar el árbol completo, (2) obtener la secuencia de alphas mediante `cost_complexity_pruning_path`, y (3) evaluar cada alpha con CV para encontrar el equilibrio óptimo.


In [30]:
path = DecisionTreeClassifier(random_state=RANDOM_STATE).cost_complexity_pruning_path(X_train, y_train)
ccp_alphas = path.ccp_alphas[:-1]  # último -> solo raíz

# Selección de alphas representativos (si hay muchos)
if len(ccp_alphas) > 12:
    idx = np.unique(np.round(np.linspace(0, len(ccp_alphas)-1, 12)).astype(int))
    alphas_sel = ccp_alphas[idx]
else:
    alphas_sel = ccp_alphas

rows = []
best = {'alpha': None, 'cv': -1}

for alpha in alphas_sel:
    tree = DecisionTreeClassifier(ccp_alpha=float(alpha), random_state=RANDOM_STATE)
    cv_scores = cross_val_score(tree, X_train, y_train, cv=5, scoring='accuracy')
    cv_mean = cv_scores.mean()

    tree.fit(X_train, y_train)
    test_acc = tree.score(X_test, y_test)

    rows.append({
        'ccp_alpha': float(alpha),
        'CV Accuracy': cv_mean,
        'Test Accuracy': test_acc,
        'Profundidad': tree.get_depth(),
        'Nº Hojas': tree.get_n_leaves()
    })
    if cv_mean > best['cv']:
        best = {'alpha': float(alpha), 'cv': cv_mean}

df_tree_ccp = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
print('Mejor alpha (en los seleccionados):', best)
df_tree_ccp


Mejor alpha (en los seleccionados): {'alpha': 0.0008896797153024911, 'cv': 0.938896985233438}


,ccp_alpha,CV Accuracy,Test Accuracy,Profundidad,Nº Hojas
2,0.000890,0.938897,0.931280,11,71
0,0.000000,0.938305,0.926540,11,97
1,0.000579,0.938305,0.928910,11,80
3,0.001120,0.935930,0.928910,10,63
4,0.001164,0.935930,0.931280,9,54
5,0.001582,0.926436,0.933649,9,47
6,0.002262,0.925846,0.938389,9,40
7,0.002772,0.919913,0.933649,8,33
8,0.004217,0.908042,0.921801,8,26
9,0.009923,0.884910,0.872038,8,19


In [31]:
# Curva alpha vs rendimiento (sobre los alphas seleccionados)
dfp = df_tree_ccp.sort_values('ccp_alpha')
fig, ax = plt.subplots(figsize=(10, 5))
ax.plot(dfp['ccp_alpha'], dfp['CV Accuracy'], marker='o', label='CV')
ax.plot(dfp['ccp_alpha'], dfp['Test Accuracy'], marker='o', label='Test')
ax.set_xlabel('ccp_alpha')
ax.set_ylabel('Accuracy')
ax.set_title('Cost-Complexity Pruning: alpha vs rendimiento')
ax.grid(True, alpha=0.3)
ax.legend()
save_fig(IMAGES_DIR / 'decision_tree_ccp_curve.png')


In [32]:
save_table(df_tree_ccp, 'tree_ccp')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_ccp.tex')

El mejor valor seleccionado es ahora **`ccp_alpha ≈ 0.00089`**, con **CV = 0.9389**, **Test = 0.9313**, profundidad **11** y **71 hojas**. La poda posterior vuelve a cumplir su papel habitual: elimina ramas poco útiles y reduce complejidad manteniendo, e incluso mejorando ligeramente, la generalización frente al árbol completo.

La tabla también muestra un comportamiento interesante: algunos alphas algo mayores reducen más la complejidad y alcanzan puntualmente valores de test incluso superiores, pero con peor CV. Por eso el mejor alpha no debe elegirse por la cifra máxima en test de una sola partición, sino por el equilibrio más estable que refleja la validación cruzada.

### 4.6 Importancia de características (mejor árbol por poda previa)

Una de las grandes ventajas de los árboles de decisión es que permiten calcular la **importancia de cada feature** según la reducción media de impureza (Gini importance). Las variables usadas en los primeros splits del árbol tienen mayor importancia.


In [33]:
best_tree = grid_tree.best_estimator_
best_tree.fit(X_train, y_train)

importances = best_tree.feature_importances_
feature_names = X_train.columns.tolist()

df_tree_imp = pd.DataFrame({
    'Feature': feature_names,
    'Importancia (Gini)': importances
}).sort_values('Importancia (Gini)', ascending=False)

df_tree_imp.head(15)


,Feature,Importancia (Gini)
3,Weight,0.491440
2,Height,0.261804
0,Gender,0.157725
5,FAVC,0.023061
1,Age,0.014733
9,CH2O,0.010856
12,TUE,0.009382
14,CAEC_Sometimes,0.008244
7,NCP,0.005938
6,FCVC,0.004786


In [34]:
# Plot top-15 importancias
topk = df_tree_imp.head(15).iloc[::-1]
fig, ax = plt.subplots(figsize=(10, 6))
ax.barh(topk['Feature'], topk['Importancia (Gini)'])
ax.set_xlabel('Importancia')
ax.set_title('Decision Tree — Importancia de características (top 15)')
save_fig(IMAGES_DIR / 'decision_tree_feature_importance.png')


In [35]:
save_table(df_tree_imp.head(20), 'tree_feature_importance')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_feature_importance.tex')

Como era de esperar, **Weight** y **Height** dominan ampliamente la importancia, coherente con lo que observamos en los scatter plots de la Práctica 0 (el IMC se calcula directamente a partir de estas dos variables). El resto de features tienen importancias mucho menores pero contribuyen colectivamente a las decisiones en niveles más profundos del árbol.

### 4.7 Visualización del árbol (profundidad limitada para legibilidad)

Visualizamos los primeros 4 niveles del mejor árbol. Se puede observar cómo las primeras divisiones utilizan Weight para separar rápidamente los niveles extremos de obesidad, y luego Height refina las categorías intermedias.

In [36]:
fig, ax = plt.subplots(figsize=(22, 10))
plot_tree(
    best_tree,
    feature_names=feature_names,
    class_names=class_names,
    filled=True,
    rounded=True,
    fontsize=8,
    max_depth=4,
    ax=ax
)
save_fig(IMAGES_DIR / 'decision_tree.png', dpi=250)

print(export_text(best_tree, feature_names=feature_names, max_depth=4))


|--- Weight <= 99.57
|   |--- Weight <= 60.06
|   |   |--- Height <= 1.66
|   |   |   |--- Weight <= 46.83
|   |   |   |   |--- Height <= 1.51
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Height >  1.51
|   |   |   |   |   |--- truncated branch of depth 5
|   |   |   |--- Weight >  46.83
|   |   |   |   |--- Height <= 1.52
|   |   |   |   |   |--- truncated branch of depth 4
|   |   |   |   |--- Height >  1.52
|   |   |   |   |   |--- truncated branch of depth 4
|   |   |--- Height >  1.66
|   |   |   |--- Weight <= 60.00
|   |   |   |   |--- TUE <= 0.02
|   |   |   |   |   |--- truncated branch of depth 2
|   |   |   |   |--- TUE >  0.02
|   |   |   |   |   |--- truncated branch of depth 6
|   |   |   |--- Weight >  60.00
|   |   |   |   |--- Height <= 1.81
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- Height >  1.81
|   |   |   |   |   |--- class: 0
|   |--- Weight >  60.06
|   |   |--- Weight <= 76.04
|   |   |   |--- Height <= 1.72
|   |   |   |   |--- FAVC <= 0.

### 4.8 Evaluación final (Parte 2)

Comparamos las tres estrategias de árboles de decisión:

- **Sin restricciones**: árbol completo sin ninguna limitación de crecimiento.
- **Mejor poda previa (pre-pruning)**: mejor combinación de `max_depth`, `min_samples_split` y `min_samples_leaf` obtenida por GridSearchCV.
- **Mejor poda posterior (post-pruning)**: mejor `ccp_alpha` seleccionado por validación cruzada.

En la tabla reportamos Test Accuracy, F1 ponderado, profundidad y número de hojas para cada variante.

In [37]:
# Mejor árbol por poda posterior (sobre los alphas seleccionados)
alpha_best = df_tree_ccp.sort_values('CV Accuracy', ascending=False).iloc[0]['ccp_alpha']
tree_ccp_best = DecisionTreeClassifier(ccp_alpha=float(alpha_best), random_state=RANDOM_STATE)

# Fit/eval
tree_ccp_best.fit(X_train, y_train)
y_pred_ccp = tree_ccp_best.predict(X_test)

tree_pre_best = grid_tree.best_estimator_
tree_pre_best.fit(X_train, y_train)
y_pred_pre = tree_pre_best.predict(X_test)

y_pred_full = tree_full.predict(X_test)

df_tree_summary = pd.DataFrame([
    {'Modelo': 'Sin restricciones', **metrics_row(y_test, y_pred_full),
     'Profundidad': tree_full.get_depth(), 'Nº Hojas': tree_full.get_n_leaves()},
    {'Modelo': 'Mejor poda previa', **metrics_row(y_test, y_pred_pre),
     'Profundidad': tree_pre_best.get_depth(), 'Nº Hojas': tree_pre_best.get_n_leaves()},
    {'Modelo': 'Mejor poda posterior', **metrics_row(y_test, y_pred_ccp),
     'Profundidad': tree_ccp_best.get_depth(), 'Nº Hojas': tree_ccp_best.get_n_leaves()},
]).sort_values('F1 (w)', ascending=False)

df_tree_summary


,Modelo,Test Accuracy,Precision (w),Recall (w),F1 (w),Profundidad,Nº Hojas
2,Mejor poda posterior,0.93128,0.934585,0.93128,0.931816,11,71
0,Sin restricciones,0.92654,0.929468,0.92654,0.927111,11,97
1,Mejor poda previa,0.92654,0.929468,0.92654,0.927111,11,97


In [38]:
save_table(df_tree_summary, 'tree_summary')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/tree_summary.tex')

In [39]:
best_tree_global = tree_pre_best  # por defecto, usamos el mejor pre-pruning
y_pred_best_tree = best_tree_global.predict(X_test)
plot_confusion(
    y_test, y_pred_best_tree, class_names,
    filename='confusion_matrix_tree.png',
    title='Matriz de Confusión — Mejor Decision Tree'
)


**Análisis de resultados (Parte 2):**

La familia de árboles cambia bastante con la nueva ejecución. Si nos limitamos a comparar baseline, poda previa y poda posterior, la **mejor poda posterior** es la que sale mejor parada con **93.13 %** de Test Accuracy, superando ligeramente al árbol sin restricciones y a la poda previa.

Sin embargo, el resultado más potente de toda esta parte está en la comparación de criterios: el árbol con **`criterion='entropy'`** o **`criterion='log_loss'`** alcanza **97.39 % en test**, muy por encima de los árboles Gini que se usan en el resto de la sección. Es decir, el verdadero hallazgo aquí no es tanto la poda como el criterio de división.

En conjunto, esto nos enseña dos cosas. La primera es que la poda posterior sigue siendo útil para compactar el modelo y rascar algo de generalización. La segunda, mucho más importante, es que en este dataset el criterio de impureza tiene un impacto enorme y puede convertir un árbol individual en un competidor de primer nivel.

---

## 5. Parte 3 — Random Forest y Extra Trees

En esta sección pasamos de un único árbol de decisión a **métodos de ensemble** basados en *bagging*. La idea central de Random Forest es entrenar múltiples árboles de decisión sobre subconjuntos aleatorios de los datos (bootstrap) y de las features, y agregar sus predicciones por votación mayoritaria. Esta diversificación reduce la varianza del modelo sin aumentar significativamente el sesgo, lo que suele mejorar la generalización respecto a un árbol individual.

Además de Random Forest, evaluaremos **Extra Trees** (Extremely Randomized Trees), una variante que introduce aún más aleatoriedad al elegir los umbrales de corte de forma aleatoria en lugar de buscar los óptimos.


---

## 5. Parte 3 — Random Forest y Extra Trees

En esta sección pasamos de un único árbol de decisión a **métodos de ensemble** basados en *bagging*. La idea central de Random Forest es entrenar múltiples árboles de decisión sobre subconjuntos aleatorios de los datos (bootstrap) y de las features, y agregar sus predicciones por votación mayoritaria. Esta diversificación reduce la varianza del modelo sin aumentar significativamente el sesgo, lo que suele mejorar la generalización respecto a un árbol individual.

Además de Random Forest, evaluaremos **Extra Trees** (Extremely Randomized Trees), una variante que introduce aún más aleatoriedad al elegir los umbrales de corte de forma aleatoria en lugar de buscar los óptimos.

### 5.1 Random Forest por defecto

Comenzamos entrenando un Random Forest con la configuración por defecto de scikit-learn (`n_estimators=100`, `max_features='sqrt'`), activando el **OOB score** (Out-of-Bag) como estimación del error de generalización sin necesidad de validación cruzada. Cada árbol se entrena sobre una muestra bootstrap (~63.2 % de los datos), y el OOB score se calcula evaluando cada muestra solo con los árboles que **no** la incluyeron en su bootstrap.

In [40]:
rf_default = RandomForestClassifier(
    n_estimators=100, random_state=RANDOM_STATE,
    n_jobs=-1, oob_score=True
)
t0 = perf_counter()
rf_default.fit(X_train, y_train)
t_fit = perf_counter() - t0

y_pred = rf_default.predict(X_test)
m = metrics_row(y_test, y_pred)

df_rf_default = pd.DataFrame([{
    'Configuración': 'RF por defecto (n=100)',
    'OOB Score': rf_default.oob_score_,
    'Test Accuracy': m['Test Accuracy'],
    'F1 (w)': m['F1 (w)'],
    'Tiempo (s)': t_fit
}])
df_rf_default


,Configuración,OOB Score,Test Accuracy,F1 (w),Tiempo (s)
0,RF por defecto (n=100),0.937129,0.945498,0.946044,0.243456


In [41]:
save_table(df_rf_default, 'rf_default')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_default.tex')

### 5.2 Efecto del número de estimadores

El parámetro `n_estimators` controla cuántos árboles se combinan en el ensemble. Más árboles proporcionan predicciones más estables (menor varianza), pero con rendimientos decrecientes y mayor coste computacional. Entrenamos Random Forest con 10, 25, 50, 100, 200 y 500 árboles para observar la curva de convergencia tanto del OOB score como de la accuracy en test.

In [42]:
rows = []
for n in [10, 25, 50, 100, 200, 500]:
    rf = RandomForestClassifier(
        n_estimators=n, random_state=RANDOM_STATE, n_jobs=-1, oob_score=True
    )
    t0 = perf_counter()
    rf.fit(X_train, y_train)
    t_fit = perf_counter() - t0

    rows.append({
        'n_estimators': n,
        'OOB Score': rf.oob_score_,
        'Test Accuracy': rf.score(X_test, y_test),
        'Tiempo (s)': t_fit
    })

df_rf_n = pd.DataFrame(rows)
df_rf_n


/Users/jordiblascolozano/Library/Python/3.11/lib/python/site-packages/sklearn/ensemble/_forest.py:611: UserWarning: Some inputs do not have OOB scores. This probably means too few trees were used to compute any reliable OOB estimates.
  warn(


,n_estimators,OOB Score,Test Accuracy,Tiempo (s)
0,10,0.858244,0.931280,0.041309
1,25,0.905101,0.933649,0.063648
2,50,0.931791,0.938389,0.111553
3,100,0.937129,0.945498,0.319793
4,200,0.943654,0.945498,0.375967
5,500,0.948992,0.950237,0.888088


In [43]:
fig, ax = plt.subplots(figsize=(10, 5))
ax.plot(df_rf_n['n_estimators'], df_rf_n['OOB Score'], marker='o', label='OOB')
ax.plot(df_rf_n['n_estimators'], df_rf_n['Test Accuracy'], marker='o', label='Test')
ax.set_xlabel('n_estimators')
ax.set_ylabel('Score')
ax.set_title('Random Forest — Efecto de n_estimators')
ax.grid(True, alpha=0.3)
ax.legend()
save_fig(IMAGES_DIR / 'rf_n_estimators_curve.png')


In [44]:
save_table(df_rf_n, 'rf_n_estimators')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_n_estimators.tex')

Se observa una mejora progresiva al aumentar el número de árboles, aunque con **rendimientos decrecientes**. Pasamos de **93.13 %** con 10 árboles a **94.55 %** con 100, y con **500 árboles** se alcanza **95.02 %**. La ganancia entre 100 y 500 árboles existe, pero ya es pequeña comparada con el incremento en tiempo de entrenamiento.

El OOB score sigue una trayectoria muy parecida y converge en torno al **94.9 %**, lo que vuelve a confirmar que es un estimador bastante fiable del error de generalización cuando el bosque ya tiene un tamaño suficiente.



### 5.3 Efecto de `max_features`

El parámetro `max_features` determina cuántas features se consideran como candidatas en cada split. Valores más bajos (`'sqrt'`, `'log2'`) aumentan la diversidad entre árboles (decorrelación), mientras que usar todas las features (`None`) permite que cada árbol sea más preciso individualmente pero aumenta la correlación entre árboles.

Comparamos cuatro opciones: `'sqrt'` ($\approx 5$), `'log2'` ($\approx 4$), `0.5` (la mitad = 11) y `None` (todas = 23).

In [45]:
rows = []
for mf in ['sqrt', 'log2', 0.5, None]:
    rf = RandomForestClassifier(
        n_estimators=100, max_features=mf,
        random_state=RANDOM_STATE, n_jobs=-1, oob_score=True
    )
    rf.fit(X_train, y_train)
    rows.append({
        'max_features': str(mf),
        'OOB Score': rf.oob_score_,
        'Test Accuracy': rf.score(X_test, y_test),
    })

df_rf_mf = pd.DataFrame(rows).sort_values('Test Accuracy', ascending=False)
df_rf_mf


,max_features,OOB Score,Test Accuracy
3,None,0.957295,0.957346
2,0.5,0.952550,0.954976
0,sqrt,0.937129,0.945498
1,log2,0.937129,0.945498


In [46]:
save_table(df_rf_mf, 'rf_max_features')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_max_features.tex')

Con la nueva ejecución, usar **todas las features** (`None`) pasa a ser la mejor opción tanto en **OOB Score** como en **Test Accuracy**: **95.73 %**. La opción `0.5` queda muy cerca (**95.50 %**), mientras que `'sqrt'` y `'log2'` se quedan claramente por detrás (**94.55 %**).

Esto refuerza la idea de que nuestro dataset tiene unas pocas variables especialmente dominantes, sobre todo **Weight** y **Height**, y conviene que estén disponibles en todos los splits importantes. Cuando se restringe demasiado el subconjunto de features candidatas, algunos árboles pierden justo la variable que más les ayudaría a separar bien las clases.



### 5.4 Grid Search (Random Forest)

Realizamos una búsqueda exhaustiva combinando `n_estimators`, `max_features`, `max_depth` y `min_samples_leaf`. Mostramos las 10 mejores configuraciones ordenadas por CV Accuracy.

In [47]:
param_grid_rf = {
    'n_estimators':     [100, 200],
    'max_features':     ['sqrt', 'log2', None],
    'max_depth':        [None, 10, 20],
    'min_samples_leaf': [1, 2, 5]
}

grid_rf = GridSearchCV(
    RandomForestClassifier(random_state=RANDOM_STATE, n_jobs=-1),
    param_grid_rf,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    refit=True
)

t0 = perf_counter()
grid_rf.fit(X_train, y_train)
t_grid = perf_counter() - t0

print("Mejores parámetros:", grid_rf.best_params_)
print(f"Mejor CV score: {grid_rf.best_score_:.4f}")
print(f"Test score: {grid_rf.best_estimator_.score(X_test, y_test):.4f}")
print(f"Tiempo GridSearch: {t_grid:.3f}s")


Mejores parámetros: {'max_depth': None, 'max_features': None, 'min_samples_leaf': 1, 'n_estimators': 100}
Mejor CV score: 0.9561
Test score: 0.9573
Tiempo GridSearch: 19.134s


In [48]:
cvres_rf = pd.DataFrame(grid_rf.cv_results_).sort_values('mean_test_score', ascending=False)
top = cvres_rf.head(10)

rows = []
for _, r in top.iterrows():
    params = {
        'n_estimators': int(r['param_n_estimators']),
        'max_features': r['param_max_features'],
        'max_depth': r['param_max_depth'],
        'min_samples_leaf': int(r['param_min_samples_leaf'])
    }
    model = RandomForestClassifier(random_state=RANDOM_STATE, n_jobs=-1, **params)
    model.fit(X_train, y_train)
    rows.append({
        'n_estimators': params['n_estimators'],
        'max_features': str(params['max_features']),
        'max_depth': 'None' if params['max_depth'] is None else int(params['max_depth']),
        'min_samples_leaf': params['min_samples_leaf'],
        'CV Acc': float(r['mean_test_score']),
        'Test Acc': model.score(X_test, y_test),
    })

df_rf_grid = pd.DataFrame(rows)
df_rf_grid


,n_estimators,max_features,max_depth,min_samples_leaf,CV Acc,Test Acc
0,100,None,None,1,0.956104,0.957346
1,100,None,20,1,0.956104,0.957346
2,200,None,None,1,0.953732,0.954976
3,200,None,20,1,0.953732,0.954976
4,200,None,10,1,0.952545,0.947867
5,100,None,10,1,0.951950,0.954976
6,200,None,20,2,0.951358,0.945498
7,200,None,None,2,0.951358,0.945498
8,200,None,10,2,0.948391,0.945498
9,100,None,20,2,0.947799,0.950237


In [49]:
save_table(df_rf_grid, 'rf_grid')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_grid.tex')

### 5.5 Importancia de características (Gini vs Permutation)

Calculamos la importancia de las features de dos maneras:

1. **Gini importance** (*Mean Decrease Impurity*): suma de las reducciones de impureza Gini ponderadas por la probabilidad de alcanzar cada nodo, promediada sobre todos los árboles del bosque. Es rápida pero puede estar **sesgada hacia features de alta cardinalidad**.
2. **Permutation importance**: mide cuánto cae la accuracy al permutar aleatoriamente los valores de una feature. Es más lenta pero más **robusta** y no tiene sesgos de cardinalidad.

Comparamos ambas métricas para las 15 features más relevantes.

In [50]:
best_rf = grid_rf.best_estimator_
best_rf.fit(X_train, y_train)

# Gini importance
gini_imp = pd.Series(best_rf.feature_importances_, index=X_train.columns)

# Permutation importance (más robusta)
perm = permutation_importance(best_rf, X_test, y_test, n_repeats=10, random_state=RANDOM_STATE, n_jobs=-1)
perm_imp = pd.Series(perm.importances_mean, index=X_train.columns)
perm_std = pd.Series(perm.importances_std, index=X_train.columns)

df_rf_imp = pd.DataFrame({
    'Feature': X_train.columns,
    'Gini Importance': gini_imp.values,
    'Permutation Importance': perm_imp.values,
    'Perm Std': perm_std.values
}).sort_values('Gini Importance', ascending=False)

df_rf_imp.head(15)


,Feature,Gini Importance,Permutation Importance,Perm Std
3,Weight,0.462227,0.732227,0.022580
2,Height,0.253635,0.323697,0.012503
0,Gender,0.158585,0.168957,0.013981
1,Age,0.029663,0.054739,0.005653
5,FAVC,0.023274,0.009242,0.003081
11,FAF,0.011333,0.001185,0.002649
12,TUE,0.009626,-0.000237,0.001659
7,NCP,0.008724,0.001185,0.003393
6,FCVC,0.008038,0.002370,0.002119
9,CH2O,0.008008,0.001185,0.001185


In [51]:
topk = df_rf_imp.head(15).iloc[::-1]
fig, ax = plt.subplots(figsize=(10, 6))
ax.barh(topk['Feature'], topk['Gini Importance'])
ax.set_xlabel('Gini Importance')
ax.set_title('Random Forest — Importancia de características (top 15)')
save_fig(IMAGES_DIR / 'rf_feature_importance_gini.png')


In [52]:
save_table(df_rf_imp.head(25), 'rf_importances')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_importances.tex')

### 5.6 Extra Trees (Extremely Randomized Trees)

**Extra Trees** difiere de Random Forest en dos aspectos clave:
- No usa *bootstrap* (se entrena con todo el dataset).
- Los **umbrales de corte se eligen al azar** en lugar de buscarse óptimamente.

Esta mayor aleatoriedad puede reducir aún más la varianza, aunque a costa de mayor sesgo. Suele funcionar bien en datasets donde Random Forest ya obtiene buenos resultados, pero no siempre mejora la generalización.

In [53]:
et = ExtraTreesClassifier(n_estimators=100, random_state=RANDOM_STATE, n_jobs=-1)

cv_et = cross_val_score(et, X_train, y_train, cv=5, scoring='accuracy')
t0 = perf_counter()
et.fit(X_train, y_train)
t_fit = perf_counter() - t0

y_pred = et.predict(X_test)
m = metrics_row(y_test, y_pred)

df_et = pd.DataFrame([{
    'Modelo': 'Extra Trees (n=100)',
    'CV Accuracy': cv_et.mean(),
    'Test Accuracy': m['Test Accuracy'],
    'F1 (w)': m['F1 (w)'],
    'Tiempo (s)': t_fit
}])
df_et


,Modelo,CV Accuracy,Test Accuracy,F1 (w),Tiempo (s)
0,Extra Trees (n=100),0.918733,0.921801,0.922909,0.126965


In [54]:
save_table(df_et, 'extra_trees')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/extra_trees.tex')

### 5.7 Evaluación final (Parte 3)

Comparamos todos los modelos basados en bagging junto con el árbol de decisión baseline como referencia. Incluimos Test Accuracy, F1 ponderado y OOB Score (cuando está disponible).

In [55]:
# Resumen comparativo Parte 3
# (Incluye baseline del árbol para referencia)

tree_baseline_test = tree_full.predict(X_test)
rf_default_pred = rf_default.predict(X_test)
rf_best_pred = best_rf.predict(X_test)
et_pred = et.predict(X_test)

df_rf_summary = pd.DataFrame([
    {'Modelo': 'Decision Tree (baseline)',
     **metrics_row(y_test, tree_baseline_test),
     'OOB Score': np.nan, 'Tiempo (s)': np.nan},
    {'Modelo': 'Random Forest por defecto',
     **metrics_row(y_test, rf_default_pred),
     'OOB Score': rf_default.oob_score_, 'Tiempo (s)': np.nan},
    {'Modelo': 'Random Forest mejor config',
     **metrics_row(y_test, rf_best_pred),
     'OOB Score': getattr(best_rf, 'oob_score_', np.nan), 'Tiempo (s)': np.nan},
    {'Modelo': 'Extra Trees',
     **metrics_row(y_test, et_pred),
     'OOB Score': np.nan, 'Tiempo (s)': np.nan},
]).sort_values('F1 (w)', ascending=False)

df_rf_summary


,Modelo,Test Accuracy,Precision (w),Recall (w),F1 (w),OOB Score,Tiempo (s)
2,Random Forest mejor config,0.957346,0.958942,0.957346,0.957597,NaN,NaN
1,Random Forest por defecto,0.945498,0.948768,0.945498,0.946044,0.937129,NaN
0,Decision Tree (baseline),0.926540,0.929468,0.926540,0.927111,NaN,NaN
3,Extra Trees,0.921801,0.925949,0.921801,0.922909,NaN,NaN


In [56]:
save_table(df_rf_summary, 'rf_summary')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/rf_summary.tex')

In [57]:
# Matriz de confusión del mejor modelo de la parte (por F1)
best_part3_name = df_rf_summary.iloc[0]['Modelo']
if best_part3_name == 'Random Forest mejor config':
    y_pred_best = rf_best_pred
elif best_part3_name == 'Random Forest por defecto':
    y_pred_best = rf_default_pred
elif best_part3_name == 'Extra Trees':
    y_pred_best = et_pred
else:
    y_pred_best = tree_baseline_test

plot_confusion(
    y_test, y_pred_best, class_names,
    filename='confusion_matrix_rf.png',
    title=f'Matriz de Confusión — Mejor modelo Parte 3: {best_part3_name}'
)


**Análisis de resultados (Parte 3):**

El **Random Forest con la mejor configuración** pasa a ser el de **100 árboles**, **todas las features** y sin límite de profundidad, alcanzando un **95.73 % de Test Accuracy** y un **F1 ponderado de 0.9576**. La mejora respecto al árbol individual baseline sigue siendo muy clara, lo que confirma el beneficio del ensemble: al agregar múltiples árboles se reduce la varianza y se suavizan errores idiosincráticos de un único árbol.

También cambia ligeramente la lectura sobre `max_features`: ahora **usar todas las variables** (`max_features=None`) no solo supera a `'sqrt'` y `'log2'`, sino también a `0.5`. Esto encaja con la estructura del problema, donde pocas variables concentran gran parte de la señal predictiva.

**Extra Trees** queda en **92.18 %**, por debajo incluso del árbol post-podado y claramente por debajo de Random Forest. La aleatorización adicional en los umbrales no parece beneficiosa aquí; introduce más sesgo del que consigue compensar con reducción de varianza.



---

## 6. Parte 4 — Gradient Boosting y AdaBoost

A diferencia del *bagging* (Random Forest), los métodos de **boosting** construyen el ensemble de forma **secuencial**: cada nuevo estimador se entrena para corregir los errores del anterior. En esta sección evaluamos dos algoritmos de boosting:

- **Gradient Boosting** (`GradientBoostingClassifier`): ajusta cada árbol al *gradiente negativo* de la función de pérdida. Permite árboles de profundidad moderada y ofrece control fino mediante `learning_rate`, `n_estimators`, `max_depth` y `subsample`.
- **AdaBoost** (`AdaBoostClassifier`): aumenta el peso de las muestras mal clasificadas en cada iteración. Clásicamente utiliza *stumps* (árboles de profundidad 1) como estimadores base.

Veremos que Gradient Boosting es especialmente potente para este dataset, mientras que AdaBoost con stumps resulta insuficiente para un problema de 7 clases.

### 6.1 Gradient Boosting por defecto

Entrenamos un `GradientBoostingClassifier` con los parámetros por defecto de scikit-learn (`n_estimators=100`, `learning_rate=0.1`, `max_depth=3`). A diferencia de Random Forest, Gradient Boosting **no admite paralelización** (los árboles se entrenan secuencialmente), por lo que el tiempo de entrenamiento será significativamente mayor.

In [58]:
gb_default = GradientBoostingClassifier(random_state=RANDOM_STATE)
cv_gb = cross_val_score(gb_default, X_train, y_train, cv=5, scoring='accuracy')

t0 = perf_counter()
gb_default.fit(X_train, y_train)
t_fit = perf_counter() - t0

y_pred = gb_default.predict(X_test)
m = metrics_row(y_test, y_pred)

df_gb_default = pd.DataFrame([{
    'Modelo': 'Gradient Boosting (default)',
    'CV Accuracy': cv_gb.mean(),
    'Test Accuracy': m['Test Accuracy'],
    'F1 (w)': m['F1 (w)'],
    'Tiempo (s)': t_fit
}])
df_gb_default


,Modelo,CV Accuracy,Test Accuracy,F1 (w),Tiempo (s)
0,Gradient Boosting (default),0.955504,0.957346,0.95792,2.058316


In [59]:
save_table(df_gb_default, 'gb_default')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/gb_default.tex')

### 6.2 Efecto de `n_estimators` y `learning_rate`

Existe un **trade-off** clásico entre estos dos parámetros: un `learning_rate` bajo necesita más estimadores para converger, pero suele generalizar mejor (cada árbol contribuye poco, evitando saltos bruscos). Un `learning_rate` alto converge rápido pero puede sobreajustar.

Probamos varias combinaciones para explorar este equilibrio.

In [60]:
configs = [
    (50,  0.1), (100, 0.1), (200, 0.1),
    (50,  0.5), (100, 0.05), (200, 0.01)
]

rows = []
for n_est, lr in configs:
    gb = GradientBoostingClassifier(
        n_estimators=n_est, learning_rate=lr, random_state=RANDOM_STATE
    )
    cv = cross_val_score(gb, X_train, y_train, cv=5, scoring='accuracy')
    gb.fit(X_train, y_train)
    rows.append({
        'n_estimators': n_est,
        'learning_rate': lr,
        'CV Accuracy': cv.mean(),
        'Test Accuracy': gb.score(X_test, y_test)
    })

df_gb_nlr = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
df_gb_nlr


,n_estimators,learning_rate,CV Accuracy,Test Accuracy
2,200,0.10,0.960251,0.973934
3,50,0.50,0.959661,0.954976
1,100,0.10,0.955504,0.957346
0,50,0.10,0.948389,0.933649
4,100,0.05,0.945422,0.936019
5,200,0.01,0.918141,0.881517


In [61]:
save_table(df_gb_nlr, 'gb_n_lr')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/gb_n_lr.tex')

La combinación **`(n_estimators=200, learning_rate=0.1)`** es ahora la mejor en **test**, alcanzando **97.39 %**, y además obtiene un **CV = 0.9603**. Muy cerca queda la configuración **`(50, 0.5)`**, que logra un CV ligeramente superior (**0.9597**) pero peor test (**95.50 %**). Este contraste es útil porque muestra que la mejor generalización en la partición de test no siempre coincide exactamente con la mejor media de validación cruzada.

Reducir el learning rate a **0.01** con solo 200 árboles vuelve a quedarse claramente corto (**88.15 %** en test): el ensemble no ha tenido suficientes iteraciones para converger. La idea general se mantiene: un learning rate muy bajo necesita muchos más árboles para resultar competitivo.



### 6.3 Efecto de `max_depth`

En Gradient Boosting, `max_depth` controla la complejidad de cada árbol individual. A diferencia de Random Forest (donde los árboles suelen ser profundos), en boosting se prefieren **árboles poco profundos** (3–5 niveles) porque el propio proceso iterativo se encarga de capturar interacciones complejas.

Árboles con `max_depth=1` (stumps) solo capturan efectos principales. Con `max_depth=3` se modelan interacciones de hasta 3 variables.

In [62]:
rows = []
for depth in [1, 2, 3, 5]:
    gb = GradientBoostingClassifier(
        max_depth=depth, n_estimators=100, learning_rate=0.1,
        random_state=RANDOM_STATE
    )
    cv = cross_val_score(gb, X_train, y_train, cv=5, scoring='accuracy')
    gb.fit(X_train, y_train)
    rows.append({
        'max_depth': depth,
        'Train Acc': gb.score(X_train, y_train),
        'CV Accuracy': cv.mean(),
        'Test Accuracy': gb.score(X_test, y_test),
    })

df_gb_depth = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
df_gb_depth


,max_depth,Train Acc,CV Accuracy,Test Accuracy
3,5,1.000000,0.964406,0.962085
2,3,0.999407,0.955504,0.957346
1,2,0.991103,0.942458,0.938389
0,1,0.882562,0.848737,0.843602


In [63]:
fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(df_gb_depth['max_depth'], df_gb_depth['Train Acc'], marker='o', label='Train')
ax.plot(df_gb_depth['max_depth'], df_gb_depth['CV Accuracy'], marker='o', label='CV')
ax.plot(df_gb_depth['max_depth'], df_gb_depth['Test Accuracy'], marker='o', label='Test')
ax.set_xlabel('max_depth')
ax.set_ylabel('Accuracy')
ax.set_title('Gradient Boosting — Efecto de max_depth')
ax.grid(True, alpha=0.3)
ax.legend()
save_fig(IMAGES_DIR / 'gb_depth_curve.png')


In [64]:
save_table(df_gb_depth, 'gb_depth')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/gb_depth.tex')

Con la nueva ejecución, **`max_depth=5`** pasa a ser la mejor opción tanto en **CV (0.9644)** como en **test (0.9621)** dentro de este barrido, por encima de `max_depth=3`. La diferencia no es enorme, pero sí suficiente para mostrar que, en este problema, árboles base un poco más profundos pueden capturar interacciones útiles sin degradar la generalización.

Aun así, el contraste con `max_depth=1` sigue siendo el mensaje principal: los stumps solo alcanzan **84.36 %**, muy lejos de las mejores configuraciones. Eso confirma que aquí se necesitan interacciones entre varias variables, no solo efectos principales aislados.



### 6.4 Grid Search (Gradient Boosting)

Combinamos `n_estimators`, `learning_rate`, `max_depth` y `subsample` en una búsqueda exhaustiva. El parámetro `subsample < 1.0` introduce *stochastic gradient boosting*, donde cada árbol se entrena con una fracción aleatoria de los datos, lo que puede mejorar la generalización y reducir el sobreajuste.

In [65]:
param_grid_gb = {
    'n_estimators':  [100, 200],
    'learning_rate': [0.01, 0.05, 0.1, 0.5],
    'max_depth':     [1, 2, 3],
    'subsample':     [0.8, 1.0]
}

grid_gb = GridSearchCV(
    GradientBoostingClassifier(random_state=RANDOM_STATE),
    param_grid_gb,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    refit=True
)

t0 = perf_counter()
grid_gb.fit(X_train, y_train)
t_grid = perf_counter() - t0

print("Mejores parámetros:", grid_gb.best_params_)
print(f"Mejor CV score: {grid_gb.best_score_:.4f}")
print(f"Test score: {grid_gb.best_estimator_.score(X_test, y_test):.4f}")
print(f"Tiempo GridSearch: {t_grid:.3f}s")


Mejores parámetros: {'learning_rate': 0.5, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.8}
Mejor CV score: 0.9609
Test score: 0.9645
Tiempo GridSearch: 63.918s


In [66]:
cvres_gb = pd.DataFrame(grid_gb.cv_results_).sort_values('mean_test_score', ascending=False).head(10)

rows = []
for _, r in cvres_gb.iterrows():
    params = {
        'n_estimators': int(r['param_n_estimators']),
        'learning_rate': float(r['param_learning_rate']),
        'max_depth': int(r['param_max_depth']),
        'subsample': float(r['param_subsample'])
    }
    model = GradientBoostingClassifier(random_state=RANDOM_STATE, **params)
    model.fit(X_train, y_train)
    rows.append({
        **params,
        'CV Acc': float(r['mean_test_score']),
        'Test Acc': model.score(X_test, y_test),
    })

df_gb_grid = pd.DataFrame(rows)
df_gb_grid


,n_estimators,learning_rate,max_depth,subsample,CV Acc,Test Acc
0,200,0.50,3,0.8,0.960850,0.964455
1,200,0.10,3,0.8,0.960253,0.962085
2,200,0.10,3,1.0,0.960251,0.973934
3,200,0.05,3,1.0,0.960251,0.952607
4,100,0.50,3,0.8,0.958476,0.962085
5,200,0.50,3,1.0,0.958475,0.957346
6,100,0.50,2,1.0,0.957883,0.954976
7,100,0.50,3,1.0,0.957881,0.964455
8,100,0.10,3,1.0,0.955504,0.957346
9,200,0.05,3,0.8,0.954912,0.945498


In [67]:
save_table(df_gb_grid, 'gb_grid')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/gb_grid.tex')

### 6.5 AdaBoost (stumps)

AdaBoost con *decision stumps* (`max_depth=1`) es un clasificador de boosting clásico. Cada stump solo puede usar **una feature y un umbral**, lo que lo convierte en un clasificador extremadamente débil. AdaBoost pondera las muestras según si fueron bien o mal clasificadas, forzando a los nuevos stumps a enfocarse en los ejemplos difíciles.

> **Nota**: Para un problema multiclase con 7 categorías, los stumps son demasiado simples como estimadores base. Esperamos resultados pobres comparados con Gradient Boosting, que usa árboles más profundos.

In [68]:
def make_adaboost(n_estimators: int, learning_rate: float) -> AdaBoostClassifier:
    """Compatibilidad sklearn: estimator (nuevo) vs base_estimator (antiguo)."""
    stump = DecisionTreeClassifier(max_depth=1, random_state=RANDOM_STATE)
    try:
        return AdaBoostClassifier(
            estimator=stump,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            random_state=RANDOM_STATE
        )
    except TypeError:
        return AdaBoostClassifier(
            base_estimator=stump,
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            random_state=RANDOM_STATE
        )

ada_configs = [
    (50,  1.0),
    (100, 1.0),
    (100, 0.5),
    (200, 0.1)
]

rows = []
for n_est, lr in ada_configs:
    ada = make_adaboost(n_est, lr)
    cv = cross_val_score(ada, X_train, y_train, cv=5, scoring='accuracy')
    t0 = perf_counter()
    ada.fit(X_train, y_train)
    t_fit = perf_counter() - t0
    rows.append({
        'n_estimators': n_est,
        'learning_rate': lr,
        'CV Accuracy': cv.mean(),
        'Test Accuracy': ada.score(X_test, y_test),
        'Tiempo (s)': t_fit
    })

df_ada = pd.DataFrame(rows).sort_values('CV Accuracy', ascending=False)
df_ada


,n_estimators,learning_rate,CV Accuracy,Test Accuracy,Tiempo (s)
3,200,0.1,0.455528,0.431280,0.529996
2,100,0.5,0.454896,0.459716,0.214931
1,100,1.0,0.448389,0.424171,0.285786
0,50,1.0,0.440721,0.535545,0.148404


In [69]:
save_table(df_ada, 'adaboost')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/adaboost.tex')

### 6.6 Evaluación final (Parte 4)

Comparamos los modelos de boosting (GB default, GB mejor configuración y AdaBoost) junto con el mejor Random Forest como referencia. Incluimos la matriz de confusión del mejor modelo de esta sección.

In [70]:
best_gb = grid_gb.best_estimator_
best_gb.fit(X_train, y_train)
y_pred_gb = best_gb.predict(X_test)

best_ada_params = df_ada.iloc[0][['n_estimators','learning_rate']].to_dict()
best_ada = make_adaboost(int(best_ada_params['n_estimators']), float(best_ada_params['learning_rate']))
best_ada.fit(X_train, y_train)
y_pred_ada = best_ada.predict(X_test)

y_pred_rf = best_rf.predict(X_test)

df_gb_summary = pd.DataFrame([
    {'Modelo': 'Random Forest (mejor config)', **metrics_row(y_test, y_pred_rf)},
    {'Modelo': 'Gradient Boosting (default)', **metrics_row(y_test, gb_default.predict(X_test))},
    {'Modelo': 'Gradient Boosting (mejor config)', **metrics_row(y_test, y_pred_gb)},
    {'Modelo': 'AdaBoost (mejor config)', **metrics_row(y_test, y_pred_ada)},
]).sort_values('F1 (w)', ascending=False)

df_gb_summary


,Modelo,Test Accuracy,Precision (w),Recall (w),F1 (w)
2,Gradient Boosting (mejor config),0.964455,0.968191,0.964455,0.964634
1,Gradient Boosting (default),0.957346,0.961186,0.957346,0.957920
0,Random Forest (mejor config),0.957346,0.958942,0.957346,0.957597
3,AdaBoost (mejor config),0.431280,0.249226,0.431280,0.311552


In [71]:
save_table(df_gb_summary, 'gb_summary')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/gb_summary.tex')

In [72]:
best_part4_name = df_gb_summary.iloc[0]['Modelo']
if best_part4_name == 'Gradient Boosting (mejor config)':
    y_pred_best = y_pred_gb
elif best_part4_name == 'AdaBoost (mejor config)':
    y_pred_best = y_pred_ada
elif best_part4_name == 'Gradient Boosting (default)':
    y_pred_best = gb_default.predict(X_test)
else:
    y_pred_best = y_pred_rf

plot_confusion(
    y_test, y_pred_best, class_names,
    filename='confusion_matrix_gb.png',
    title=f'Matriz de Confusión — Mejor modelo Parte 4: {best_part4_name}'
)


**Análisis de resultados (Parte 4):**

**Gradient Boosting** sigue siendo la familia más fuerte del bloque, pero la interpretación ahora es algo más sutil. El **GridSearchCV** selecciona como mejor configuración por CV a **`learning_rate=0.5`, `max_depth=3`, `n_estimators=200`, `subsample=0.8`**, con **CV = 0.9609** y **Test = 0.9645**. Sin embargo, en la propia tabla del grid aparece una configuración casi idéntica en CV, **`learning_rate=0.1`, `max_depth=3`, `n_estimators=200`, `subsample=1.0`**, que alcanza un **Test = 0.9739**, el mejor valor observado en esta familia.

Esto significa que, con la nueva ejecución, el boosting sigue dominando, pero la diferencia entre seleccionar por CV y observar el mejor test concreto se vuelve especialmente visible. Metodológicamente, el notebook mantiene como `best_estimator_` la configuración ganadora por validación cruzada, que es la decisión correcta; aun así, merece la pena dejar constancia de que otra combinación del mismo grid generaliza mejor en esta partición concreta.

**AdaBoost con stumps** vuelve a mostrar un rendimiento muy pobre. Según el criterio de CV, la mejor configuración se queda en torno a **45.6 %** de accuracy y **43.1 %** en test; según test puro, el caso de 50 estimadores llega a **53.55 %**, pero sigue siendo muy insuficiente para competir con el resto de modelos. El problema sigue siendo el mismo: los stumps son demasiado débiles para un problema multiclase de este nivel.



---

## 7. Comparación global de mejores modelos

Finalmente, seleccionamos el **mejor modelo de cada parte** y los enfrentamos en una comparación directa. Para cada modelo calculamos CV Accuracy (5-fold), Test Accuracy, Precision, Recall, F1 ponderado y tiempo de entrenamiento. Esto nos permite evaluar no solo la calidad predictiva sino también el coste computacional de cada enfoque.

Los modelos seleccionados son:
- **SVM (best RBF)**: Pipeline con StandardScaler + SVC(kernel='rbf', C=100, gamma=0.01)
- **Decision Tree (best pre-pruning)**: mejor configuración por GridSearchCV
- **Random Forest (best)**: 100 árboles, todas las features, sin límite de profundidad
- **Gradient Boosting (best)**: configuración seleccionada por `GridSearchCV`, es decir, `learning_rate=0.5`, `max_depth=3`, `n_estimators=200`, `subsample=0.8`
- **AdaBoost (best)**: incluido para contrastar su bajo rendimiento con stumps


In [73]:
# Seleccionamos mejores modelos por bloque
best_models = {
    'SVM (best RBF)': grid_rbf.best_estimator_,
    'Decision Tree (best pre-pruning)': grid_tree.best_estimator_,
    'Random Forest (best)': grid_rf.best_estimator_,
    'Gradient Boosting (best)': grid_gb.best_estimator_,
    'AdaBoost (best)': best_ada,
}

rows = []
for name, model in best_models.items():
    # CV accuracy
    cv = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
    cv_mean, cv_std = cv.mean(), cv.std()

    # Fit time
    t0 = perf_counter()
    model.fit(X_train, y_train)
    fit_time = perf_counter() - t0

    y_pred = model.predict(X_test)
    m = metrics_row(y_test, y_pred)

    rows.append({
        'Modelo': name,
        'CV Accuracy': cv_mean,
        'CV Std': cv_std,
        **m,
        'Tiempo (s)': fit_time
    })

df_final = pd.DataFrame(rows).sort_values('F1 (w)', ascending=False)
df_final


,Modelo,CV Accuracy,CV Std,Test Accuracy,Precision (w),Recall (w),F1 (w),Tiempo (s)
3,Gradient Boosting (best),0.960850,0.009472,0.964455,0.968191,0.964455,0.964634,3.293068
2,Random Forest (best),0.956104,0.007851,0.957346,0.958942,0.957346,0.957597,0.186398
0,SVM (best RBF),0.940087,0.008094,0.940758,0.941686,0.940758,0.940521,0.099791
1,Decision Tree (best pre-pruning),0.938305,0.010725,0.926540,0.929468,0.926540,0.927111,0.009128
4,AdaBoost (best),0.455528,0.012941,0.431280,0.249226,0.431280,0.311552,0.487621


In [74]:
save_table(df_final[['Modelo','CV Accuracy','Test Accuracy','Precision (w)','Recall (w)','F1 (w)','Tiempo (s)']], 'final_comparison')

PosixPath('/Users/jordiblascolozano/Documents/IA/3erAnyo/2doCuatri/aprendizaje_avanzado/practicas/practica_1/tables/final_comparison.tex')

In [75]:
best_name = df_final.iloc[0]['Modelo']
best_model = best_models[best_name]
best_model.fit(X_train, y_train)
y_pred_best = best_model.predict(X_test)

print('Mejor modelo global:', best_name)
print(classification_report(y_test, y_pred_best, target_names=class_names))

plot_confusion(
    y_test, y_pred_best, class_names,
    filename='confusion_matrix_final.png',
    title=f'Matriz de Confusión — Mejor modelo global: {best_name}'
)


Mejor modelo global: Gradient Boosting (best)
                     precision    recall  f1-score   support

Insufficient_Weight       1.00      0.91      0.95        54
      Normal_Weight       0.85      1.00      0.92        57
     Obesity_Type_I       0.99      0.97      0.98        70
    Obesity_Type_II       0.97      1.00      0.98        60
   Obesity_Type_III       1.00      1.00      1.00        65
 Overweight_Level_I       1.00      0.88      0.94        58
Overweight_Level_II       0.97      0.98      0.97        58

           accuracy                           0.96       422
          macro avg       0.97      0.96      0.96       422
       weighted avg       0.97      0.96      0.96       422



**Análisis global y conclusiones:**

La comparativa final generada por el notebook confirma que **Gradient Boosting** sigue siendo el mejor modelo entre los seleccionados automáticamente por cada bloque, con **96.45 % de Test Accuracy** y **F1 ponderado = 0.9646**. Le siguen **Random Forest** con **95.73 %**, **SVM (RBF)** con **94.08 %**, **Decision Tree (best pre-pruning)** con **92.65 %** y finalmente **AdaBoost** con un rendimiento muy pobre (**43.13 %**).

Observaciones clave:

1. **La complejidad del ensemble sigue marcando la diferencia**: los mejores ensembles superan con claridad a las versiones más simples del árbol y a AdaBoost con stumps.
2. **SVM continúa siendo competitivo**, pero tras eliminar el leakage el kernel lineal resulta incluso más sólido que el RBF optimizado. El problema parece más lineal de lo que sugería la ejecución anterior.
3. **El caso de árboles merece una lectura crítica**: aunque la comparativa global usa el árbol seleccionado por GridSearch, la tabla de criterios muestra que `entropy` y `log_loss` alcanzan **97.39 % en test**, un resultado extraordinario para un árbol individual.
4. **Gradient Boosting ofrece dos lecturas complementarias**: el mejor modelo por CV logra **96.45 %**, pero una configuración vecina del grid (`n_estimators=200`, `learning_rate=0.1`, `max_depth=3`, `subsample=1.0`) llega a **97.39 %** en test.
5. **AdaBoost con stumps no es viable** para este problema multiclase complejo. Incluso su mejor caso práctico queda muy lejos del resto.
6. **Trade-off tiempo/rendimiento**: Random Forest mantiene un equilibrio especialmente bueno entre calidad y coste, mientras que Gradient Boosting tarda más pero sigue liderando entre los modelos seleccionados automáticamente.

